# Project 12 - Anthony Micco

**TA Help:** N/A
    
**Collaboration:** N/A

## Question 1

In [1]:
#defining the polars code to see how fast it runs
import polars as pl
import time
start_time = time.time()
myDF = pl.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv", null_values="NA")
print("My program took", time.time() - start_time, "to run")

My program took 2.361281394958496 to run


In [2]:
#defining the pandas code to see how fast it runs
import pandas as pd
import time
start_time = time.time()
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv")
print("My program took", time.time() - start_time, "to run")

My program took 5.779568672180176 to run


Polars runs about 5-6 times as fast as pandas does when loading this data set as Polars took aproximately 1.05 seconds to run, while Pandas took aproximately 6.1 seconds to run.

## Question 2

In [3]:
# calculating the time it takes to read in the dataset in Polars
import polars as pl
# read in the 2002 flight data to a Polars data frame again
myDF = pl.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv", null_values="NA")

import time
start_time = time.time()
myresults = myDF.filter(pl.col("Origin") == "IND")['DepDelay'].mean()
print("My program took", time.time() - start_time, "to run")
print("The average delay for Indianapolis flights was " + str(myresults))

My program took 0.0636148452758789 to run
The average delay for Indianapolis flights was 3.9418510691198407


In [4]:
# calculating the time it takes to read in the dataset in Pandas
import pandas as pd
# read in the 2002 flight data to a Pandas data frame again
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv")

import time
start_time = time.time()
myresults = myDF[myDF['Origin'] == "IND"]['DepDelay'].mean()
print("My program took", time.time() - start_time, "to run")
print("The average delay for Indianapolis flights was " + str(myresults))

My program took 0.29450297355651855 to run
The average delay for Indianapolis flights was 3.9418510691198407


Polars runs about 10 times faster than Pandas as the Polars code ran in aproximately 0.025 seconds while the Pandas code ran in aproximately 0.28 seconds.

## Question 3

In [5]:
# calculating the time it takes to get average flight delay for all California flights using Polars
import polars as pl
# read in the 2002 flight data to a Polars data frame again, but only 3 columns this time:
myDF = pl.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv", null_values="NA").select(["Origin", "Dest", "DepDelay"])

# also read in the airports data to a Polars data frame as follows:
myairports = pl.read_csv("/anvil/projects/tdm/data/flights/subset/airports.csv")

import time
start_time = time.time()
newDF = myDF.join(myairports, left_on="Origin", right_on="iata")
myresults = newDF.filter(pl.col("state") == "CA")['DepDelay'].mean()
print("My program took", time.time() - start_time, "to run")
print("The average delay for all flights with Origin at any California airports was " + str(myresults))

My program took 0.4099109172821045 to run
The average delay for all flights with Origin at any California airports was 5.245392521079324


In [6]:
# calculating the time it takes to get average flight delay for all California flights using Pandas
import pandas as pd
# read in the 2002 flight data to a Pandas data frame again, but only 3 columns this time:
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv", usecols=["Origin", "Dest", "DepDelay"])

# also read in the airports data to a Pandas data frame as follows:
myairports = pd.read_csv("/anvil/projects/tdm/data/flights/subset/airports.csv")

import time
start_time = time.time()
newDF = myDF.merge(myairports, left_on='Origin', right_on='iata')
myresults = newDF[newDF['state'] == "CA"]['DepDelay'].mean()
print("My program took", time.time() - start_time, "to run")
print("The average delay for Indianapolis flights was " + str(myresults))

My program took 1.283733606338501 to run
The average delay for Indianapolis flights was 5.245392521079324


Polars runs about 3-4 times faster than Pandas at running this program because Polars completed it in aproximately 0.35 seconds and Pandas completed it in aproximately 1.35 seconds.

## Question 4

In [7]:
# calculating the time it takes to find the number of unique Orgin airports using Polars
# read in the 2002 flight data to a Polars data frame
myDF = pl.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv", null_values="NA")

import time
start_time = time.time()
#getting the number of unique origins
numOrigins = len(myDF["Origin"].unique())

#creating a list of each of the unique origins
uniqueOrigins = myDF["Origin"].unique()

print("My program took", time.time() - start_time, "to run")

#printing out the num and names of the unique origins 
print("The number of unique origin airports is " + str(numOrigins))
print("The unique origins are: ")
for origin in uniqueOrigins:
    print(origin)

My program took 0.1313610076904297 to run
The number of unique origin airports is 218
The unique origins are: 
LAN
BHM
ILM
FAR
DTW
BRW
PIT
SJU
GRR
DUT
IND
PDX
BET
CAK
SYR
BPT
CHA
BDL
TUL
CLL
DRO
TXK
MIA
ERI
MFE
FLL
STL
DBQ
TUS
PHX
OKC
PSC
PSP
MLU
RNO
KTN
XNA
EUG
SGF
CDV
WRG
OAK
LSE
MYR
MEM
LAS
SMF
OGG
KOA
EWR
LBB
DAY
BIS
BGR
GFK
CMH
LGB
ANC
LAW
BOI
CLE
ISP
AGS
ELP
EVV
JNU
MFR
CLT
GEG
SCC
SFO
BIL
MOB
AMA
SBA
MLI
EGE
TPA
STT
SNA
FWA
BRO
MTJ
GUC
SBP
RSW
TOL
MSN
MAZ
BNA
FCA
ATL
TLH
MLB
ADQ
AZO
AVL
JAN
TYR
MBS
ABE
CAE
GPT
MAF
ILE
HDN
DFW
JFK
SAT
RST
LEX
MDT
BMI
ORD
LIT
AKN
DEN
HNL
MRY
SBN
GSP
COS
RDU
MQT
SEA
SAN
DLG
PSG
MOT
IAH
LRD
MSO
ABI
ALB
BTV
PNS
AUS
MCI
MCO
ORF
DAL
LAX
OTZ
MSP
GRB
LIH
EYW
SIT
OME
DAB
CRP
STX
CID
SUX
CVG
AVP
YAK
SWF
FNT
SLC
PVD
BQN
HPN
GST
ACT
HSV
PWM
LFT
RAP
DSM
MHT
SAV
CHS
GGG
IAD
MKE
ICT
ONT
FAI
BUF
FSM
JAX
SJC
PIA
SRQ
VPS
SPS
PBI
FAT
RIC
MSY
SDF
TVC
DLH
TYS
LGA
OMA
PHL
MDW
ROC
GTF
BWI
BTR
CMI
HOU
ORH
SHV
ABQ
SJT
DCA
FSD
BOS
BUR
GSO
JAC
HRL
HLN
BZN


In [8]:
# calculating the time it takes to find the number of unique Orgin airports using Pandas
# read in the 2002 flight data to a Pandas data frame
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv")

import time
start_time = time.time()
#getting the number of unique origins
numOrigins = len(myDF["Origin"].unique())

#creating a list of each of the unique origins
uniqueOrigins = myDF["Origin"].unique()

print("My program took", time.time() - start_time, "to run")

#printing out the num and names of the unique origins 
print("The number of unique origin airports is " + str(numOrigins))
print("The unique origins are: ")
for origin in uniqueOrigins:
    print(origin)

My program took 0.3075685501098633 to run
The number of unique origin airports is 218
The unique origins are: 
PIT
PHL
STL
CLT
MHT
BNA
ROC
SFO
GSO
RDU
IAH
SAV
BWI
DCA
PBI
DTW
MCI
BOS
AVP
ORD
IND
CLE
ORF
MCO
TPA
BUF
SJU
MSY
ATL
FLL
EWR
MYR
MKE
BDL
MSP
RIC
MDT
SDF
BHM
CMH
PVD
CHS
ALB
PNS
IAD
PWM
LGA
AVL
MIA
DFW
MEM
RSW
DAY
ILM
SYR
SAN
BTV
ERI
JAX
SEA
HPN
STT
STX
DAL
HOU
LBB
LIT
CRP
TUL
HRL
MAF
AMA
OKC
ABQ
ELP
SMF
LAX
PHX
AUS
MDW
SLC
LAS
ONT
SJC
ISP
PDX
JAN
RNO
BUR
OAK
BOI
TUS
SAT
GEG
SNA
OMA
BGR
JFK
EYW
GRR
MSN
FSD
FAR
BIS
MBS
COS
TVC
RAP
VPS
RST
ABE
GRB
DEN
DLH
BZN
FCA
MOT
GFK
FNT
LAN
GTF
ANC
CVG
HNL
BIL
SRQ
GPT
MSO
DSM
LSE
BTR
PSP
AZO
GSP
HDN
EGE
LIH
OGG
KOA
ICT
CID
EUG
MFR
JAC
GUC
MOB
LEX
AGS
HSV
MLU
MLB
SHV
HLN
DAB
TLH
TYS
CAE
PSC
CAK
LGB
SBP
SBA
FAT
MRY
ABI
ACT
TXK
FSM
ILE
LAW
LRD
BPT
GGG
LFT
SGF
SJT
XNA
SPS
CLL
FWA
TYR
DRO
MAZ
ORH
CMI
BMI
PIA
EVV
DBQ
TOL
MQT
CHA
MFE
SWF
MTJ
MLI
DUT
ADQ
BET
OME
OTZ
SCC
JNU
KTN
CDV
YAK
SIT
PSG
WRG
FAI
AKN
DLG
BRW
BQN
BRO
SBN
GST
SUX


The Polars program ran about 2-3 times faster than the Pandas program since the Polars program ran in aproximately 0.23 seconds and the Pandas program ran in aproximately 0.32 seconds. 

## Question 5

In [9]:
#reading in the data set to see what the columns are
myDF = pl.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv", null_values="NA")
myDF.head()

shape: (5, 29)
┌──────┬───────┬────────────┬───────────┬───┬──────────────┬──────────┬──────────────┬─────────────┐
│ Year ┆ Month ┆ DayofMonth ┆ DayOfWeek ┆ … ┆ WeatherDelay ┆ NASDelay ┆ SecurityDela ┆ LateAircraf │
│ ---  ┆ ---   ┆ ---        ┆ ---       ┆   ┆ ---          ┆ ---      ┆ y            ┆ tDelay      │
│ i64  ┆ i64   ┆ i64        ┆ i64       ┆   ┆ str          ┆ str      ┆ ---          ┆ ---         │
│      ┆       ┆            ┆           ┆   ┆              ┆          ┆ str          ┆ str         │
╞══════╪═══════╪════════════╪═══════════╪═══╪══════════════╪══════════╪══════════════╪═════════════╡
│ 2002 ┆ 1     ┆ 13         ┆ 7         ┆ … ┆ null         ┆ null     ┆ null         ┆ null        │
│ 2002 ┆ 1     ┆ 14         ┆ 1         ┆ … ┆ null         ┆ null     ┆ null         ┆ null        │
│ 2002 ┆ 1     ┆ 15         ┆ 2         ┆ … ┆ null         ┆ null     ┆ null         ┆ null        │
│ 2002 ┆ 1     ┆ 16         ┆ 3         ┆ … ┆ null         ┆ null     ┆ null         ┆ null        │
│ 2002 ┆ 1     ┆ 17         ┆ 4         ┆ … ┆ null         ┆ null     ┆ null         ┆ null        │
└──────┴───────┴────────────┴───────────┴───┴──────────────┴──────────┴──────────────┴─────────────┘

In [10]:
#using polars to determine the average airtime for flights with a destination of PIT
myDF = pl.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv", null_values="NA")

import time
start_time = time.time()

#finding the average air time for flights to Pittsburgh
airtime = myDF.filter(pl.col("Dest") == "PIT")['AirTime'].mean()
print("My program took", time.time() - start_time, "to run")
print("The average airtime for flights to Pittsburgh was " + str(airtime))

My program took 0.02689051628112793 to run
The average airtime for flights to Pittsburgh was 88.75755032506169


In [11]:
#using pandas to determine the average airtime for flights with a destination of PIT
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/subset/2002.csv")

import time
start_time = time.time()

#finding the average air time for flights to Pittsburgh
airtime = myDF[myDF["Dest"] == "PIT"]["AirTime"].mean()
print("My program took", time.time() - start_time, "to run")
print("The average airtime for flights to Pittsburgh was " + str(airtime))

My program took 0.3055088520050049 to run
The average airtime for flights to Pittsburgh was 88.75755032506169


The Polars program was about 10 times faster than the Pandas program because the Polars program ran in about 0.027 seconds while the Pandas program ran in about 0.31 seconds. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.